### DISCUSSION TITLE
arduino_driver makes the program stop

### DISCUSSION THREAD

**Discussion by IETVE** [tdslite/tdslite#28] [Q&A] [answered]
Hi,

My program stops as soon as it reaches this line : `tdsl::arduino_driver<EthernetClient> driver {net_buf};`

I tested with the `01-initialized-library.ino` example, same problem. 

So I moved the driver declaration below `Serial.Begin()` and `Ethernet.begin()`, then did some `Serial.println()` and pinged the IP I gave to the arduino. 
As long as the program doesn't reach the driver declaration line, the Serial Monitor displays correctly and the Ping responds. Until the program gets to this problematic line, it stops and nothing responds.

So, it's definitely this line.

_I'm using the Portenta Machine Control, containing a Portenta H7._

Code : 
```C
#include <Ethernet.h>
#include <tdslite.h>
... 
...

byte mac [] = {0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xEF}; 
IPAddress ip(192, 168, 0, 45);

void setup() {
  Serial.begin(9600);
  while (!Serial);
  Serial.println("Serial Initialized");

  Ethernet.begin(mac, ip);
  Serial.println("Ethernet Initialized");

  Serial.println("Waiting for Ping");
  delay(5000);

  tdsl::uint8_t net_buf [768] = {};
  Serial.println("net_buf OK !");

  tdsl::arduino_driver<EthernetClient> driver {net_buf};
  Serial.println("Driver OK !");

  decltype(driver)::progmem_connection_parameters params;
  Serial.println("Params instantiated");
  ...
  ...

}
```

Results : 

![image_2023-08-14_100040428](https://github.com/mustafakemalgilor/tdslite/assets/142207871/1c93e2f2-be92-4240-8faa-3a488760f256)

![image_2023-08-14_101059420](https://github.com/mustafakemalgilor/tdslite/assets/142207871/8fd4f16b-13f7-418a-9147-7ca43b373277)

**Comment by xmkg** [comment]
Hello @IETVE;

First of all, the driver object is stack allocated in your example, which means it'll be destroyed as soon as the "setup()" method finishes, which is not probably what you want. You should move it to the global scope.

As for the problem itself, it's quite interesting that that line is making your program freeze. What that line is doing in the background is as follows:

- Forward constructor arguments to tdsl::tds_context 
```c++
        template <typename... Args>
        inline tdsl_driver(Args &&... args) noexcept : tds_ctx(TDSL_FORWARD(args)...) {}
```

- Call "NetworkImplementation"'s constructor (which is tdsl::tdsl_netimpl_arduino) 
```c++
        template <typename... Args>
        inline tds_context(Args &&... args) noexcept :
            NetworkImplementation(TDSL_FORWARD(args)...) {
            // Register tds_context message handler to network implementation
            this->register_packet_data_callback(&handle_packet_data, this);
        }
```
- ... which in turn default-constructs the "EthernetClientType"' (which is EthernetClient)
```c++
        template <typename Z>
        inline tdsl_netimpl_arduino(tdsl::byte_span network_io_buffer, EthernetClientType ec = {}) :
            client(TDSL_MOVE(ec)) {
            this->network_buffer = tdsl::tdsl_buffer_object{network_io_buffer};
        }
```

So we basically default construct an "EthernetClient" in the end. Can you test whether this freezes as well:

```c++
#include <Ethernet.h>
#include <tdslite.h>
... 
...

byte mac [] = {0xDE, 0xAD, 0xBE, 0xEF, 0xFE, 0xEF}; 
IPAddress ip(192, 168, 0, 45);

void setup() {
  Serial.begin(9600);
  while (!Serial);
  Serial.println("Serial Initialized");

  Ethernet.begin(mac, ip);
  Serial.println("Ethernet Initialized");

  Serial.println("Waiting for Ping");
  delay(5000);

  tdsl::uint8_t net_buf [768] = {};
  Serial.println("net_buf OK !");

  EthernetClient ec  = {}; //  previously was tdsl::arduino_driver<EthernetClient> driver {net_buf};
  Serial.println("Driver OK !");

  decltype(driver)::progmem_connection_parameters params;
  Serial.println("Params instantiated");
  ...
  ...

}

```

Let me know how that goes.

**Reply by IETVE** [reply]
Thank you for your responsiveness.

`driver` and `net_buf` are in `Setup()` for test purposes only.

With `EthernetClient ec = {};`, `driver` is no longer declared and therefore cannot be used in `decltype(driver)`. 
This creates compilation errors.

What can I use in this case?

**Reply by xmkg** [reply]
Hello@IETVE, just comment out the parameters(progmem_connection_parameters) as well. We're trying to see whether the problem is caused by the EthernetClient's constructor or not.

**Reply by IETVE** [reply]
`EthernetClient ec  = {};` doesn't block the program.

**Reply by xmkg** [reply]
OK, thanks for testing. Can you define the following macros before the `#include <tdslite.h>`:

```C++

/**
 * Print a formatted string to serial output.
 *
 * @param [in] FMTSTR Format string
 * @param [in] ...    Format arguments
 *
 * The format string will be stored in program memory in order
 * to save space.
 */
#define SERIAL_PRINTF(FMTSTR, ...)                                                                 \
    [&]() {                                                                                        \
        /* Save format string into program */                                                      \
        /* memory to save flash space */                                                           \
        static const char __fmtpm [] PROGMEM = FMTSTR;                                             \
        char buf [64]                        = {};                                                 \
        snprintf_P(buf, sizeof(buf), __fmtpm, ##__VA_ARGS__);                                      \
        Serial.print(buf);                                                                         \
    }()

#define SERIAL_PRINTLNF(FMTSTR, ...)                                                               \
    SERIAL_PRINTF(FMTSTR, ##__VA_ARGS__);                                                          \
    Serial.println("");                                                                            \
    Serial.flush()

#define SERIAL_PRINT_U16_AS_MB(U16SPAN)                                                            \
    [](tdsl::u16char_view v) {                                                                     \
        for (const auto ch : v) {                                                                  \
            Serial.print(static_cast<char>(ch));                                                   \
        }                                                                                          \
    }(U16SPAN)

#define TDSL_DEBUG_PRINT(FMTSTR, ...)       SERIAL_PRINTF(FMTSTR, ##__VA_ARGS__)
#define TDSL_DEBUG_PRINTLN(FMTSTR, ...)     SERIAL_PRINTLNF(FMTSTR, ##__VA_ARGS__)
#define TDSL_DEBUG_PRINT_U16_AS_MB(U16SPAN) SERIAL_PRINT_U16_AS_MB(U16SPAN)
```


... and also can you put a `Serial.flush()` after the "Driver OK" message to see if that changes anything:

```C++
  tdsl::arduino_driver<EthernetClient> driver {net_buf};
  Serial.println("Driver OK !");
  Serial.flush();
```

Are you calling the "driver.connect()" method after initializing the driver? I'm asking this because it would make more sense if the code is stuck at "connect()" than the driver's constructor.

**Reply by IETVE** [reply]
No change. It stays on **net_buf OK!** and doesn't flush.

Something strange, the arduino seems to disconnect after `tdsl::arduino_driver<EthernetClient> driver {net_buf};` 
_(That explains why nothing is written on the serial monitor)._

A popup error message says: `Port monitor error: command 'open' failed: Serial port not found. Could not connect to COM4 serial port`

`driver.connect(params);` is at the end after the SQL server data auth. 

Like this: 

```C
  tdsl::uint8_t net_buf [768] = {};
  Serial.println("net_buf OK !");

  tdsl::arduino_driver<EthernetClient> driver {net_buf};
  Serial.println("Driver OK !");

  decltype(driver)::progmem_connection_parameters params;
  Serial.println("Params instantiated");

  params.server_name = TDSL_PMEMSTR(SRV_IP);
  Serial.println("IP set");

  params.user_name = TDSL_PMEMSTR(SRV_USER);
  Serial.println("User set");

  params.password = TDSL_PMEMSTR(SRV_PASSWORD);
  Serial.println("Password set");

  params.packet_size = { 512 };
  Serial.println("Packet size set");

  driver.connect(params);
  Serial.println("DB Connected");
```

**Reply by xmkg** [reply]
Interesting. I have three hypotheses right now:

- Given that the board has an ARM-based STM32H747XI MCU, the board might be crashing due to an illegal instruction (I haven't tested the code on ARM yet)
- a timer/watchdog is resetting the board (maybe due to some exception/error)
- board might be going oom somehow (unlikely)

I've ordered a Portenta H7 + Vision Shield combo, I'll expect it to arrive in 1 week so I can better diagnose the issue when it arrives. Meanwhile, I'll try to reproduce it in a virtual environment (e.g. QEMU emulated ARM) to see whether the issue is specific to ARM or not.

Can you also try reproducing the issue with the onboard WiFi as well?

**Reply by IETVE** [reply]
Same goes for WiFi.

I've tried different IDEs but nothing changes.
Next, I'll try to debug step by step. VS22 can do it, I think.

**Reply by NicolasInterHM** [reply]
Hi. I’ve got the same problem with a Finder Opta AFX00002.

**Reply by xmkg** [reply]
Hi @NicolasInterHM, thanks for reporting the issue.

It seems like Opta also uses the STM32H747XI MCU (the same MCU as Portenta H7), so it's most likely an ARM-specific issue. I ordered a Portenta H7 last week and it's now arrived in customs -- which means I'll probably have the means to dig into this issue further tomorrow (or Friday).

I'll update this discussion as I progress.

**Reply by xmkg** [reply]
Hi folks,

Today, I got my hands on Portenta H7 and found a weird Arduino bug instead of a tdslite one. It seems like mbed EthernetClient implementation has an invalid move constructor, so moving an EthernetClient to another crashes the board:

Simple reproducer:

```c++
#include <Ethernet.h>

// --------------------------------------------------------------------------------
// remove_reference

template<typename T>
struct remove_reference {
  typedef T type;
};

template<typename T>
struct remove_reference<T &> {
  typedef T type;
};

template<typename T>
struct remove_reference<T &&> {
  typedef T type;
};


#define MOVE(...) \
  static_cast<typename remove_reference<decltype(__VA_ARGS__)>::type &&>( \
    __VA_ARGS__)

void setup() {
  EthernetClient ec = {};
  EthernetClient ec2{ MOVE(ec) };  // Crashes the board
}

void loop() {}
```

I opened a ticket in ArduinoCore-mbed https://github.com/arduino/ArduinoCore-mbed/issues/715 to get it fixed. Meanwhile, I'll issue a temporary fix for the mbed boards.

**Reply by xmkg** [reply, marked as answer]
@IETVE, @NicolasInterHM 

The issue should be fixed with the [0.4.0 release](https://github.com/mustafakemalgilor/tdslite/releases/tag/v0.4.0); let me know if you experience issues after the update.

**Reply by IETVE** [reply]
You rocks !!!

Thank you so much for everything you've done.
Everything works now.

**Comment by NicolasInterHM** [comment]
Awesome, thanks for the update! Everything works for me as well.

Le ven. 25 août 2023 à 10:05, Igor Etienne ***@***.***> a
écrit :
